### ISSUE TITLE
MPI async collectives and tile lifetime

### ISSUE THREAD

**Issue by albestro** [eth-cscs/DLA-Future#368]
As they are implemented now, `MPI_Ireduce` and `MPI_Iallreduce` do an additional internal copy in case the memory being used for the operation is not contiguous.

This is a workaround we opted for because of the impossibility to use predefined `MPI_Op` with `MPI_Vector` (used by our tiles when the memory is not contiguous). AFAIK, the preset of `MPI_Op` (e.g. MPI_SUM) can work just with contiguous memory, while for `MPI_Vector` it is possible to define custom `MPI_Op` using the [`MPI_Op_create`](https://www.mpich.org/static/docs/latest/www3/MPI_Op_create.html).

In async collectives with copy workaround, we have to deal with different phases:
- make the original tile contiguous (if not yet)
- use the contiguous data for the asynchronous operation (being it the original tile or the temporary buffer)
- in case a tile has been used for output of the collective operation, and the temporary buffer was used, copy back the data to the original tile

Across all these phases, the original tile must be kept alive until the asynchronous operations is completed (due to MPI requirement).

For the moment I started investigating a possible solution with the workaround in branch [alby/test/reduce_async_mech](https://github.com/eth-cscs/DLA-Future/tree/alby/test/reduce_async_mech) (see [test/unit/communication/test_comm_matrix.cpp](https://github.com/eth-cscs/DLA-Future/blob/dc735c8b04dcb94d2b46c2745479db346c9c8e69/test/unit/communication/test_comm_matrix.cpp))

It does not look like very nice, and I'm wondering if it wouldn't be better to fix this workaround by implementing a custom `MPI_Op` for `MPI_Vector`. Probably it will not fix all our problems (I'm thinking about the CPU/GPU interaction), but it may improve the context a bit.

Looking forward to your feedbacks!

@msimberg @rasolca @teonnik

**Comment by albestro** [comment]
This morning I started investigating the custom `MPI_Op` for `MPI_Vector` solution for trying to fix the workaround.

The main problem I'm facing is that we create `MPI_Datatype` at runtime, in particular we create a datatype `MPI_Vector` with the layout information `(cols, rows, ld)`. This tuple is not known at compile time since it depends on the matrix construction.

This is problematic because the new `MPI_Op` implementation has to be specified to `MPI_Op_create` by passing a C function-pointer to a function with a signature like this

```c
void MPI_User_function(void *invec, void *inoutvec, int *len, MPI_Datatype *datatype);
```

AFAIK, it is not possible to query the `datatype` about its structure, and the remaining option I thought about is keeping a mapping between the custom datatype and its operator, so that in the function we can assume as precondition that we know for sure that the datatype is what we think it is. But, the problem here it is that the implementation has to be "stateful", since it has to depend on the tuple with the layout given by the matching datatype, so it cannot be a simple C-function (since we cannot add a "context" parameter).

Beside this problem for which I'm not able to think about a solution with current constraints, I don't know how much different in terms of performance a custom reduce operation implementation on non-contiguous data ("reduce non-cont") may be w.r.t. an operation on a contiguous buffer ("copy to contiguous + reduce cont")

note: IIRC currently the datatype we create represents the full tile, so  the `len` argument equals the number of tiles we are sending (i.e. 1 if we are sending a single tile).

**Comment by msimberg** [comment]
> This morning I started investigating the custom `MPI_Op` for `MPI_Vector` solution for trying to fix the workaround.
> 
> The main problem I'm facing is that we create `MPI_Datatype` at runtime, in particular we create a datatype `MPI_Vector` with the layout information `(cols, rows, ld)`. This tuple is not known at compile time since it depends on the matrix construction.
> 
> This is problematic because the new `MPI_Op` implementation has to be specified to `MPI_Op_create` by passing a C function-pointer to a function with a signature like this
> 
> ```c
> void MPI_User_function(void *invec, void *inoutvec, int *len, MPI_Datatype *datatype);
> ```
> 
> AFAIK, it is not possible to query the `datatype` about its structure, and the remaining option I thought about is keeping a mapping between the custom datatype and its operator, so that in the function we can assume as precondition that we know for sure that the datatype is what we think it is. But, the problem here it is that the implementation has to be "stateful", since it has to depend on the tuple with the layout given by the matching datatype, so it cannot be a simple C-function (since we cannot add a "context" parameter).
> 
> Beside this problem for which I'm not able to think about a solution with current constraints, I don't know how much different in terms of performance a custom reduce operation implementation on non-contiguous data ("reduce non-cont") may be w.r.t. an operation on a contiguous buffer ("copy to contiguous + reduce cont")
> 
> note: IIRC currently the datatype we create represents the full tile, so the `len` argument equals the number of tiles we are sending (i.e. 1 if we are sending a single tile).

Take this with a grain of salt as I know almost nothing about MPI. What exactly is the `datatype` argument? Since you provide the user function, can't you cast `datatype` to whatever type you know it was? If the type can be different the user function can also be templated.

As for the changes in your branch, I have the feeling it could be simplified. One thing to check is if it would help to turn that input future of a tile into a shared future of a tile. It might make managing the data going between those steps a bit simpler. Also, remember that there's `getUnwrapReturnValue` to help with accessing just the return value if you're using `unwrapExtendTiles`: https://github.com/eth-cscs/DLA-Future/blob/87c6e12b45ebc1f73ed5ce89cd72d19d5bb0a969/include/dlaf/matrix/tile.h#L309-L316.

**Comment by albestro** [comment]
> Take this with a grain of salt as I know almost nothing about MPI. What exactly is the `datatype` argument? Since you provide the user function, can't you cast `datatype` to whatever type you know it was? If the type can be different the user function can also be templated.

Unfortunately, `MPI_Datatype` is just an handle to the type inside MPI, so it is just an alias for an `int`.

> As for the changes in your branch, I have the feeling it could be simplified. One thing to check is if it would help to turn that input future of a tile into a shared future of a tile. It might make managing the data going between those steps a bit simpler. 

I'm not really sure what future/shared_future you are targeting, but I'm with you that there may be a chance to improve the current version of the code.

> Also, remember that there's `getUnwrapReturnValue` to help with accessing just the return value if you're using `unwrapExtendTiles`:

I was aware of its existence, but for the purpose I have to access both the return value and a specific argument (i.e. the tile that keeps the ownership and I want to keep alive also for the next steps).
As soon as I realize what are the main use-cases, I will try to factor out helpers like (e.g. maybe something like `getUnwrapReturnAndArgs`).

**Comment by msimberg** [comment]
> > Take this with a grain of salt as I know almost nothing about MPI. What exactly is the `datatype` argument? Since you provide the user function, can't you cast `datatype` to whatever type you know it was? If the type can be different the user function can also be templated.
> 
> Unfortunately, `MPI_Datatype` is just an handle to the type inside MPI, so it is just an alias for an `int`.

Ah, it doesn't actually contain any data, it's literally just the type of the data? If that's the case, is `invec` or `inoutvec` something you know the type of?

> 
> > As for the changes in your branch, I have the feeling it could be simplified. One thing to check is if it would help to turn that input future of a tile into a shared future of a tile. It might make managing the data going between those steps a bit simpler.
> 
> I'm not really sure what future/shared_future you are targeting, but I'm with you that there may be a chance to improve the current version of the code.

I was mainly thinking about the one passed into `scheduleX`.

> > Also, remember that there's `getUnwrapReturnValue` to help with accessing just the return value if you're using `unwrapExtendTiles`:
> 
> I was aware of its existence, but for the purpose I have to access both the return value and a specific argument (i.e. the tile that keeps the ownership and I want to keep alive also for the next steps).
> As soon as I realize what are the main use-cases, I will try to factor out helpers like (e.g. maybe something like `getUnwrapReturnAndArgs`).

Might it be clearer in this case to just return the specific argument instead of relying on `unwrapExtendTiles` to return it? A combination of `unwrapping` and explicitly returning what you need to return might make it clearer?

**Comment by albestro** [comment]
> > > Take this with a grain of salt as I know almost nothing about MPI. What exactly is the `datatype` argument? Since you provide the user function, can't you cast `datatype` to whatever type you know it was? If the type can be different the user function can also be templated.
> > 
> > 
> > Unfortunately, `MPI_Datatype` is just an handle to the type inside MPI, so it is just an alias for an `int`.
> 
> Ah, it doesn't actually contain any data, it's literally just the type of the data? If that's the case, is `invec` or `inoutvec` something you know the type of?

I'm not an MPI expert too, but AFAIK if you are implementing a primitive type operation (e.g. datatype == MPI_FLOAT) then you can easily check that the handler is the same and then you can infer what is the type of the invec/inoutvec.

In our case, extending what I was saying before, we may decide to keep the match between a `MPI_Vector` and its specialized `MPI_Op` for its tuple `(rows, cols, ld)` (e.g. with a global map `<MPI_Datatype, MPI_Op>`), that would allow to use the correct `MPI_Op` when we call a collective for a specific `MPI_Datatype`

e.g. `MPI_Ireduce(..., message.data(), dlaf::get_op(message.type()), message.type())`

But, even if this allow would allow us to know in advance the `MPI_Datatype` for a specific custom `MPI_Op`, I still se the problem of accessing the tuple with the configuration of the datatype, which is known at runtime.

While I'm writing this I started thinking about the global state solution, so we can make the custom operation point to a global map where also this information are stored...it doesn't sound very elegant, efficient and clean anyway.

**Comment by albestro** [comment]
## TL;DR
Copying a non-contiguous to a contiguous memory does not seem to be just a workaround, but the right thing to do.

So, I will pursue working on the original "HPX solution" with async tasks. I will open a PR with a seed of the proposal soon.

## Report
A few days ago with @rasolca, we did a couple of performance tests about "contiguous vs non-contiguous MPI communications".

Each test run has 3 phases:
1. Contiguous: using contiguous memory, mainly as a reference for comparison
2. Strided: use MPI with non-contiguous memory
3. Strided with Contiguous temporary: a non-contiguous memory is copied to a contiguous memory, communicated (i.e. contiguous communication), and then copied back to the original non-contiguous (not interesting for a BCast, but it is for Reduce).

In the specific cases I'm reporting here, "tiles" are 256x256 `double` with a leading dimension of 10240 elements for the strided case.

*note1*:  In order to avoid some binding problems leading to artifacts in the results, we forced the binding with this line

```
srun -n 2 numactl -N 0 --membind=0 ./perf_test
```

## Broadcast 2Nodes

The first test we tried to replicate was one for which @rasolca already collected some results about it: `MPI_Bcast` between 2 nodes.

Each column reports the time elapsed (ns) for the activity (0# is the result from the rank 0). In the latter case, there are three additional columns reporting pre-copy/communication/post-copy timings.

```
CONT        STRIDED      STRIDED with TEMP
0#5539271   0#1485036    0#339781        192285 93020 54476
0#64260     0#1424443    0#164772        50797 59750 54225
0#60904     0#169036     0#163807        50524 59022 54261
0#60560     0#182363     0#163895        50752 59190 53953
0#60650     0#167411     0#163482        50081 59238 54163
0#60448     0#167447     0#164129        50630 59109 54390
0#61031     0#190432     0#163549        50346 59242 53961
0#60432     0#167957     0#175582        50760 70516 54306
0#60406     0#167639     0#164061        50808 59209 54044
0#60648     0#167049     0#163918        50676 59135 54107
0#60657     0#184914     0#164111        50468 59128 54515
0#60709     0#167222     0#163256        50365 58833 54058
0#60982     0#167620     0#164082        50726 59142 54214
```

Considerations (valid for this specific configuration):
- Contiguous communication takes 60us;
- Strided communication takes 170us (~3x)
- Strided with temp: the communication takes as expected ~60us, and together with pre/post copy we reach the Strided communication performance (MPI probably does the same thing) 

## Reduce

Following previous comments about `MPI_Op_create`, @rasolca identified a chance in MPI for "querying" the datatype, that lead to this naive implementation for a custom MPI_SUM operator compatible with MPI_Vector.

```cpp
template <class T>
void op_vector_impl(void *invec, void *inoutvec, int *len, MPI_Datatype *datatype) {
  int nblocks, blocksize, stride;
  MPI_Datatype element_type;
  std::tie(nblocks, blocksize, stride, element_type) = query_vector(*datatype);
  
  assert(*len == 1);
  
  for (int i_b = 0; i_b < nblocks; ++i_b) {
    for (int i_e = 0; i_e < blocksize; ++i_e) {
      const int i = i_b * stride + i_e;
      static_cast<T*>(inoutvec)[i] += static_cast<T*>(invec)[i];
    }
  }
}
```

### Reduce 2Nodes

```
CONT        STRIDED      STRIDED with TEMP
0#3233728   0#2170059    0#1083415       264240 764365 54810
0#3509030   0#2540745    0#247218        49186 145593 52439
0#155315    0#1045407    0#245568        48949 144337 52282
0#153099    0#1039606    0#246047        48824 144740 52483
0#154299    0#1101492    0#245810        48999 144237 52574
0#154417    0#1063401    0#245071        48794 144037 52240
0#153498    0#1040425    0#245833        48979 144649 52205
0#153391    0#1038740    0#245090        48865 143762 52463
0#155816    0#1101109    0#245510        48649 144474 52387
0#154027    0#1043250    0#245729        48928 144270 52531
0#153979    0#1038520    0#261646        48880 158166 54600
0#153776    0#1037934    0#246547        49085 144733 52729
0#154055    0#1080685    0#250542        50566 147480 52496
```

Considerations:
- Contiguous MPI in first and last example performances ~match (remember: in the last the communication happens over the temporary contiguous)
- The pre-post copy of the tile is confirmed to cost ~50us
- The strided reduce with the naive operator is very expensive ~4x

### Reduce 2Nodes (NOP)

Just as additional info, reported here a test where the for loop in the custom operator has been commented out (just the type query is performed)

```
CONT        STRIDED     STRIDED with TEMP
0#3262282   0#1968431   0#1054047       259241 739053 55753
0#3396358   0#2380035   0#244797        49754 138472 56571
0#147089    0#882076    0#239795        49377 136953 53465
0#145336    0#871343    0#256279        65659 136919 53701
0#146528    0#903023    0#239945        49589 136714 53642
0#141623    0#898659    0#239769        49292 136639 53838
0#145729    0#870233    0#239118        49368 136120 53630
0#145402    0#867488    0#238763        49430 135972 53361
0#145577    0#868450    0#239723        49561 136788 53374
0#155030    0#868182    0#239944        49443 136982 53519
0#173567    0#881770    0#239972        49230 137282 53460
0#145713    0#884545    0#239778        49421 136602 53755
0#146509    0#871970    0#260984        49218 136180 75586
```

Considerations:
- Strided reduce, even without performing an operation, is way too expensive ~3.5x
- Other timings are more or less confirmed

### Reduce 4Nodes

As last test, the naive implementation with a reduce over 4 nodes.

```
CONT        STRIDED       STRIDED with TEMP
0#3329776   0#3049832     0#1127386       272030 805077 50279
0#3549553   0#2805820     0#309094        47012 212098 49984
0#235883    0#2125061     0#304379        46865 207629 49885
0#249119    0#2391390     0#301916        46672 205177 50067
0#276644    0#2490961     0#334745        46841 209260 78644
0#244674    0#2131077     0#305778        46739 209122 49917
0#247003    0#2123267     0#341024        79372 211656 49996
0#244599    0#2149791     0#304266        46844 207347 50075
0#230147    0#2120877     0#308292        46749 211587 49956
0#248188    0#2120552     0#306959        46779 210169 50011
0#241423    0#2123095     0#312358        46983 215520 49855
0#240125    0#2124472     0#314076        46875 217136 50065
0#236220    0#2123538     0#314248        46820 217388 50040
```

Considerations:
- Contiguous MPI communication in first and last phase ~match
- Tile copy always costs ~50us
- Strided MPI is very expensive ~6x
